apps/emqx/src/emqx_persistent_session_ds_inflight.erl

```diff
@@ -129,12 +158,191 @@ n_buffered(all, #inflight{n_qos0 = NQos0, n_qos1 = NQos1, n_qos2 = NQos2}) ->
 n_inflight(#inflight{n_inflight = NInflight}) ->
     NInflight.
 
+-spec puback(emqx_persistent_session_ds:seqno(), t()) -> {ok, t()} | {error, Expected} when
+    Expected :: emqx_persistent_session_ds:seqno() | undefined.
+puback(SeqNo, Rec = #inflight{puback_queue = Q0, n_inflight = N}) ->
+    case ipop(Q0) of
+        {{value, SeqNo}, Q} ->
+            {ok, Rec#inflight{
+                puback_queue = Q,
+                n_inflight = max(0, N - 1)
+            }};
+        {{value, Expected}, _} ->
+            {error, Expected};
+        _ ->
+            {error, undefined}
+    end.
+
+-spec pubcomp(emqx_persistent_session_ds:seqno(), t()) -> {ok, t()} | {error, Expected} when
+    Expected :: emqx_persistent_session_ds:seqno() | undefined.
+pubcomp(SeqNo, Rec = #inflight{pubcomp_queue = Q0, n_inflight = N}) ->
+    case ipop(Q0) of
+        {{value, SeqNo}, Q} ->
+            {ok, Rec#inflight{
+                pubcomp_queue = Q,
+                n_inflight = max(0, N - 1)
+            }};
+        {{value, Expected}, _} ->
+            {error, Expected};
+        _ ->
+            {error, undefined}
+    end.
+
+%% PUBREC doesn't affect inflight window:
```

Comment:
Currently, we seem to [ignore](https://github.com/emqx/emqx/blob/811edb32a28d3948b59e37f29baaaa6c05e5828b/apps/emqx/src/emqx_channel.erl#L438) the PUBREC reason code, but it seems it _might_ influence the inflight window:

> The send quota is incremented by 1:
> 
> ·         Each time a PUBACK or PUBCOMP packet is received, regardless of whether the PUBACK or PUBCOMP carried an error code.
> 
> ·         Each time a PUBREC packet is received with a Return Code of 0x80 or greater.

Reply:
Indeed: https://github.com/emqx/emqx/blob/755b59b7fe21870e4d000387bda8bce1c91cf8a4/apps/emqx/src/emqx_channel.erl#L438

I'd prefer to fix it in a different PR, since it affects MEM sessions as well.